monash-human-power/MHP_DAS
# Automated Data Logging

## Description
These changes allow for automated data logging into an SQLite database each time a true `v3/start` message is received and subsequent conversion into an excel file when a false message is received. The excel files are stored in the `dashboard/server/data` folder to be uploaded onto Google Drive.

Updates have also been made to the relevant `README` and `env` files.

## Screenshots
_Screenshot 1_
![image](https://user-images.githubusercontent.com/97724307/213292808-5e68b8a7-b419-4f43-8231-d5c7250701c9.png)

_Screenshot 2_
![image](https://user-images.githubusercontent.com/97724307/213292914-491addcd-6232-466f-a63c-b2f4860af00d.png)

_Screenshot 3_
![image](https://user-images.githubusercontent.com/97724307/213293133-72f25ba7-aa3d-4818-8567-238b3d9985ef.png)

_Screenshot 4_
![image](https://user-images.githubusercontent.com/97724307/213293286-ebf40c19-f587-4627-a859-bdc9ffc0e4a5.png)

_Screenshot 5_
![image](https://user-images.githubusercontent.com/97724307/213293868-b03a6b58-d0fc-4adb-b060-e53a2bcfe3c7.png)

_Screenshot 6_
![image](https://user-images.githubusercontent.com/97724307/213294003-4f7a2cc4-e47c-44c5-b028-9d4c2d475eb7.png)


## Steps to Test

1. In a terminal navigate to the correct directory to `data-acquisition-system/DAS/das`.
2. Set up your `.env` file accordingly.
    - `MQTT_LOG_FILE` should be your local path that leads to the `data-acquisition-system/DAS/das/sqlite/` folder
    - `EXCEL_LOG_FILE` should be your local path that leads to the `dashboard/server/data/` folder.
3. Spawn a poetry shell using `poetry env use 3.7`.
    - **Note if this doesn't work use `poetry env use PATH_TO_PY3.7`.**
4. Open the poetry shell and run `poetry install`.
5. Open a split terminal navigating to `data-acquisition-system/DAS/das`, and open a poetry shell.
6. In the first terminal run `python log_2_excel.py`, you may use any command line arguments but take note of the `host`.
7. In the second terminal run `python V3_fake_module.py`.
   - You should notice that a true start message is received and an SQLite database should appear in the `sqlite` folder, as per _Screenshots 1 & 2_.
8. Stop the V3_fake_module by running `CTRL + C` in that terminal.
   - A false start message should be received, and an excel file should be saved in the correct folder, as per _Screenshots 3 & 4_.
   - Check that the data has been correctly logged into the excel file by opening it
9. To test that we can only start if the logger hasn't started and stop when the logger has started follow the below:
   - In a terminal run `python log_2_excel.py`.
   - Send a false start message, `mosquitto_pub -t "v3/start" -m '{"start": false}'`, and check the terminal shows something similar to _Screenshot 5_.
   - Send a true start message, `mosquitto_pub -t "v3/start" -m '{"start": true}'`, and then send it again, check the terminal shows something similar to _Screenshot 6_.

For extra testing, try sending messages using different brokers such as the one on the primary display. Testing can also be done by switching to this branch on the primary display and running the code on there to check that logs are saved correctly.
